### DISCUSSION TITLE
UI Design

### DISCUSSION THREAD

**Discussion by luc-github** [luc-github/ESP3D-TFT#6] [Ideas]
The official UI Editor of lvgl has already a sample UI for 3D printer : https://squareline.io/
as it is a sample / a proof of concept it has several issues to be used :
1 - it is designed for a 800x600 screen so it use complex UI which will not fit most of `cheap` screen in current hardware target list
2 - it only use images => which make the UI fancy but also huge and which definitly will not fit the ESP32 FS 
3 - It cover only half of the needed UI => many settings are not covered : file selection, wifi configuration with keyboard usage, multiextruders, CNC UI etc...

But it bring several ideas on what controls should be used and some of the user flow
1 - need a main menu : print job information / jog / settings / TBD
2 - Use roller control for value selection is definitly a must
3 - slider control need + / - additionnal controls IMHO to allow precise changes
4 - we need to limit the usage of images and use the fonts as much as possible to limit the final size https://docs.lvgl.io/master/overview/font.html#special-fonts
But need adding some additionnal specific fonts may be necessary, I did it before (https://github.com/luc-github/svg-icons/tree/master/networkfonts) when network font where not available - now it seems not necessary for network but for specific features like extruder / bed / heat / laser / etc... it may still be a good vs an embedded image TBD

5 - Use LVG buttons / switches / rollers / labels / lists / keyboard should cover al controls
6 - fancy UI like jog one in sample  or webUI should be simple to keep ease of use (avoid fat finger issue) ,  reliablity (need to do what it is supposed to do) and size (keep low foot print), so they are not a priority I think for the moment
7 - try to use relative positions as mucha as possible for controls to keep UI consistent across resolution or we will have to define a absolute positions file which may be painful to maintain at the end each time there is UI change
8 - theme support in lvgl is enabled but it is not a priority - first need to define / design each pages - theming will be in next steps once UI is clearly defined
9 - unlike webUI, displaying printer EEPROM settings is not a target for the screen UI - at least at this moment, it will be more part of next steps

if anyone has some design skill please shoot I am not designer and not good a it 😓 

Any suggestion / mockup / of UI and user flow is welcome

**Comment by Skorpi08** [comment]
https://dribbble.com/shots/13953444-Touch-UI

![image](https://user-images.githubusercontent.com/5186502/193410149-8ab7d82e-b8ea-4f14-990e-21918cc25969.png)

**Comment by luc-github** [comment]
For reference here all controls/widgets available  in lvgl, of course style can be modified : 
Animated image:
![animateImage](https://user-images.githubusercontent.com/8822552/193411810-7c6604ef-98ef-4d43-9c14-ba3c9fd8f454.png)

Arc:
![arc](https://user-images.githubusercontent.com/8822552/193411813-f380d532-3b25-48e3-8a10-9b7116afb3c6.png)
![arc2](https://user-images.githubusercontent.com/8822552/193411814-76ae3d34-2ee0-4ed1-932d-1acceaca9cd9.png)

Bar:
![bar](https://user-images.githubusercontent.com/8822552/193411816-eac88346-c07a-4d0c-80d5-6c41186dc89f.png)
![bar2](https://user-images.githubusercontent.com/8822552/193411817-4ab669e1-80f9-4d1d-8600-4b5f31741457.png)
![bar3](https://user-images.githubusercontent.com/8822552/193411818-34e3225a-3367-4c9f-b9b0-5ae2b0b021e8.png)
![bar4](https://user-images.githubusercontent.com/8822552/193411819-8d7c103b-0a35-4a31-8a72-5940bc0bb86b.png)
![bar5](https://user-images.githubusercontent.com/8822552/193411820-2a7aaf60-85d4-4eaa-9c78-ef0d34ac1523.png)

Button:
![button](https://user-images.githubusercontent.com/8822552/193411821-5fcce222-d4a1-41e2-825c-a9853bc8b667.png)

Button Matrix:
![buttonMatrix](https://user-images.githubusercontent.com/8822552/193411823-8aacb442-c11a-4b1e-98e3-b6520f8bb85d.png)
![buttonmatrix2](https://user-images.githubusercontent.com/8822552/193411825-adeadb37-005d-4c4a-8bc9-73ecc36e858f.png)
![buttonmatrix3](https://user-images.githubusercontent.com/8822552/193411826-e5adca20-2068-46fb-b9fa-b4ebcaaa82b4.png)

Calendar:
![calendar](https://user-images.githubusercontent.com/8822552/193411828-7e2754f8-90b3-4429-8c87-ff2367f2fdbe.png)

Canvas:
![canvas](https://user-images.githubusercontent.com/8822552/193411830-6de2fbf9-e203-4cc7-8b3f-f284dbfebb42.png)

Chart:
![chart](https://user-images.githubusercontent.com/8822552/193411831-95a3fd97-de78-4443-879c-a0686aa415c5.png)
![chart2](https://user-images.githubusercontent.com/8822552/193411832-c81025c7-1479-4294-8cd3-97abd09b2d97.png)
![chart3](https://user-images.githubusercontent.com/8822552/193411833-db73d9ff-669f-4eee-9e27-06d76425ce10.png)
![chart4](https://user-images.githubusercontent.com/8822552/193411834-b45edf30-0686-46e5-95d9-34942bfe74a6.png)
![chart5](https://user-images.githubusercontent.com/8822552/193411835-183adc4e-4e2e-4b69-b00c-14d322f7bb17.png)
![chart6](https://user-images.githubusercontent.com/8822552/193411836-056ae4a8-2788-4799-a352-bdd154516a19.png)
![chart7](https://user-images.githubusercontent.com/8822552/193411837-0a116518-6639-49d1-81e7-2b532951f82f.png)

Checkbox:
![checkbox](https://user-images.githubusercontent.com/8822552/193411838-8e0ac5f0-4fbe-4363-86ff-6326b9f082ef.png)
![checkbox2](https://user-images.githubusercontent.com/8822552/193411839-732feea9-c436-44b2-9381-b26583929829.png)

Color wheel:
![colorwheel](https://user-images.githubusercontent.com/8822552/193411840-5a75071b-bbd1-46fe-8014-8dfcbc7f4020.png)

Dorplist:
![droplist](https://user-images.githubusercontent.com/8822552/193411841-d22d3ab8-40a0-45bd-bfc9-d22d1bf982ac.png)

Images:
![image](https://user-images.githubusercontent.com/8822552/193411842-e0ed391a-47e2-4034-a8ec-aa25590c0a3c.png)

Image button:
![imagebutton](https://user-images.githubusercontent.com/8822552/193411844-3082d5c4-4c4d-4783-81ef-6d36725100b9.png)

Keyboard:
![keyboard](https://user-images.githubusercontent.com/8822552/193411845-9cf5df8b-e5bc-4a49-a7c3-7069a5ae46fd.png)

Label:
![label](https://user-images.githubusercontent.com/8822552/193411846-d0757d23-a973-48cc-9213-419b9005c3c9.png)

Led:
![led](https://user-images.githubusercontent.com/8822552/193411849-574ee9b8-d172-465d-a7da-b11764aae544.png)

Lines:
![lines](https://user-images.githubusercontent.com/8822552/193411850-5ecc4202-ccc9-442f-8d09-efc95829b86b.png)

List:
![list](https://user-images.githubusercontent.com/8822552/193411851-8a9e4f13-06fc-4fa2-9564-892dd2484f59.png)
![list2](https://user-images.githubusercontent.com/8822552/193411852-f88f20f9-47d9-460d-8e76-522d3e0320fe.png)

Menu:
![menu](https://user-images.githubusercontent.com/8822552/193411853-2d768849-fdf5-4361-9952-18f16e193e4b.png)
![menu2](https://user-images.githubusercontent.com/8822552/193411854-48b5c0e1-b505-42b0-8482-439112c8efdb.png)
![menu3](https://user-images.githubusercontent.com/8822552/193411855-07e8d403-33bc-4533-98a2-e134b01f31b6.png)
![menu4](https://user-images.githubusercontent.com/8822552/193411856-4204835b-687d-4cfd-9832-59b50dac1ea3.png)

Message Box:
![messagebox](https://user-images.githubusercontent.com/8822552/193411857-3ae86e35-8e77-490f-b329-244634bdc2ed.png)

Meter:
![meter](https://user-images.githubusercontent.com/8822552/193411859-ea0f655d-c1fe-4c70-bcca-95fc759abc56.png)
![meter2](https://user-images.githubusercontent.com/8822552/193411861-9358f99e-51a1-451d-abb4-41b969353b67.png)
![meter3](https://user-images.githubusercontent.com/8822552/193411863-cf45b65f-4e07-4d23-81b8-f8c77f441823.png)
![meter4](https://user-images.githubusercontent.com/8822552/193411865-5a543a5d-45e4-41a3-a513-06992e242ad4.png)

Roller:
![roller](https://user-images.githubusercontent.com/8822552/193411866-d930ff61-c365-4e38-a817-17c6c7333ada.png)
![roller2](https://user-images.githubusercontent.com/8822552/193411867-dbd6f434-a49e-486c-857e-b0da8eb6ae7a.png)

Slider:
![slider](https://user-images.githubusercontent.com/8822552/193411868-831c69bd-bb39-4c9b-9ffd-df7d62abfbc8.png)
![slider2](https://user-images.githubusercontent.com/8822552/193411869-eb951e2c-e868-4662-a513-19ec99e4741b.png)
![slider3](https://user-images.githubusercontent.com/8822552/193411870-f1dd190e-4f0b-4af1-b45f-fa63ecbc26a6.png)
![span](https://user-images.githubusercontent.com/8822552/193411871-5fcf21b0-83d9-4fcd-bc8d-d55ed0a101c7.png)

Spinbox:
![spinbox](https://user-images.githubusercontent.com/8822552/193411873-33209a5a-98f0-464b-a5ff-ecc0d9f04d88.png)

Spinner:
![spinner](https://user-images.githubusercontent.com/8822552/193411874-a2d5dde3-aed7-4480-9739-74155f0efa29.png)

Switch:
![switch](https://user-images.githubusercontent.com/8822552/193411875-df82bcbd-1470-43ff-8b50-4f166a9ab913.png)

Table:
![table](https://user-images.githubusercontent.com/8822552/193411876-acb69e75-00c6-4f1f-94d3-5667fb0170df.png)
![table2](https://user-images.githubusercontent.com/8822552/193411877-d323200b-76c1-4c45-b221-c4a2a6562c11.png)

Table view:
![tableview](https://user-images.githubusercontent.com/8822552/193411879-ca1d116e-ca39-41b5-8a48-c57535e81769.png)
![tableview2](https://user-images.githubusercontent.com/8822552/193411880-efb26a8e-4b24-4ed8-9b82-d2c986f3caf2.png)

Text area:
![textarea](https://user-images.githubusercontent.com/8822552/193411881-1c1ee1df-1650-44e1-b9b2-d7da389f0ed9.png)
![textarea2](https://user-images.githubusercontent.com/8822552/193411882-7a174123-054a-44be-bb43-17c4828de329.png)
![textarea3](https://user-images.githubusercontent.com/8822552/193411883-626c37dc-702e-4dc2-8f6a-8905cf2e015f.png)

Titleview:
![titleview](https://user-images.githubusercontent.com/8822552/193411884-2c59033e-07c8-407c-906b-51aba0cb0da7.png)

Window:
![window](https://user-images.githubusercontent.com/8822552/193411886-eb9986d3-be33-4644-a7d9-6d6a2149e31e.png)

**Reply by Skorpi08** [reply]
it should be mentioned, not all are available with the UI Editor.

**Reply by 3DSmitty** [reply]
Here is Marlins LCD Menu Tree which could be used for reference:

https://marlinfw.org/docs/features/lcd_menu.html

**Reply by luc-github** [reply]
I have started to draw some UI org - followed some inkscape training to do the necessary context icons- but did not work on real mockup yet - still busy with backend code 
I will start when wifi is ready to see the impact of performances when wifi and tft are active at once

**Comment by luc-github** [comment]
the current target screen have different resolutions:
![screens](https://user-images.githubusercontent.com/8822552/228417566-fb887fc1-88f5-473b-830b-83873a42b713.png)

but there real size are not really different:  2.8` , 3.5' and 4.3' 
![proportions](https://user-images.githubusercontent.com/8822552/228417841-6e4adb24-1b40-4e2a-8025-61b7423bafa0.png)

because there is no 7' in target size I suggest to focus in unique UI that will fit all.

Here my proposal for 3d printer :
![image](https://user-images.githubusercontent.com/8822552/228421480-fbb88468-46b5-4928-8309-3b4c41fe9734.png)
it cover 4 extruders + 1 heat bed max:
heat bed report will alternate with chamber report if no space
sensors are only output and will also alternate if over number of 3
during print settings / Print / TBD / buttons will be hidden
Progression will appear in status bar, eventually pushing status bar may display more informations (TBD)
TBD button would be the Jog button 
I do not think display position in main screen is really a big added value but it is arguable , it could be a button in same level as extruders that display X/Y/Z and would be then the jog button or babystep button when printing=>TBD

Using some [icons / symboles](https://docs.lvgl.io/latest/en/html/overview/font.html) would avoid to use any images: 
![image](https://user-images.githubusercontent.com/8822552/228423004-b253efe9-cb7b-43c1-bfbe-2707adfb2b66.png)
IMHO there are 10 symboles, at least, that can be overwriten for specific usages:
extruder / bed / chamber / probe or sensor / fan / speed etc..

Feedback are welcome if you have better suggestion
the other screens need to be created but they should follow same idea

Sub screens also could look like previous suggestions: https://github.com/luc-github/ESP3D-TFT/discussions/6#discussioncomment-3777932 because they have less constraint of display : https://github.com/luc-github/ESP3D-TFT/discussions/6#discussioncomment-3777932

so  

For GRBL it should be defined what need to be displayed :
Status bar also for all text message
Machine Positions: up to 8 which open jog screen
Working positions: up to 8
in that case only display one kind of position at once? add toogle button on screen to choose which one to see ?
State (IDLE) running/job
Pause/resume button
Reset button
probe button ?
Settings
file selection button 

Any fedback is welcome

**Reply by 3DSmitty** [reply]
@luc-github Looks Awesome. Question / Comments about the status bar. Will this show serial output from printer board? Something I like about the btt lcd's is in the status box it shows any serial messages from main board. Also somewhere on the main screen it should show the XYZ current location, will this be displayed in status bar as well?

**Reply by Skorpi08** [reply]
for Serial output an extra menu would be better I think

**Reply by luc-github** [reply]
@3DSmitty  BTT does not show all serial messages - it only show M117 XXX / echo: XXX / error: XXXX
So behavior should be same

**Reply by luc-github** [reply]
@Skorpi08  yes I was also thinking to allow to display last 10 update in another screen

**Reply by 3DSmitty** [reply]
@luc-github Yes that is what I meant M117 messages, perfect.

**Reply by 3DSmitty** [reply]
@luc-github How about M154 to report current XYZ location. Will that be in status bar as well?

**Reply by luc-github** [reply]
no
I wrote 
>I do not think display position in main screen is really a big added value but it is arguable , it could be a button in same level as extruders that display X/Y/Z and would be then the jog

**Reply by 3DSmitty** [reply]
@luc-github  ok, well it should absolutely be shown when a print job is in progress screen then. And on the jog / movement screen.

**Reply by luc-github** [reply]
Jog and printing are different usage, in Jog it is mandatory of course
But when printing - what is the purpose ? 
Monitoring temperature is a must - monitoring position when printing what is the benefit ? I am wondering

**Reply by 3DSmitty** [reply]
@luc-github I always monitor the Z height value when printing, this tells me a few things. Guestimation of how much time is left, at what height a certain feature of the part is going to print (overhangs, supports, critical features) that I may want to monitor a little more closely. I do a lot of prototyping so having the ability to numerically know what is coming on the next layer is very important.

**Reply by luc-github** [reply]
then as I wrote :
> it could be a button in same level as extruders that display X/Y/Z and would be then the jog button

**Reply by 3DSmitty** [reply]
@luc-github ok as long I can access it during a print.

**Reply by 3DSmitty** [reply]
@luc-github Idea for the TBD button. How about a during print button for things like babystepping and fine tuning?

**Reply by luc-github** [reply]
the babysteps should be jog button - why add another one ?

**Reply by 3DSmitty** [reply]
@luc-github ok

**Comment by luc-github** [comment]
Here the first draft of possible symbols from free [fontawesome](https://fontawesome.com/search?o=r&m=free)
I currently put together cnc and 3d printing but can be splited if necessary, I only used the solid of version 6.4.0.
![image](https://user-images.githubusercontent.com/8822552/229337134-c271714e-c494-45eb-9d1c-5c92468f26c0.png)

Need to regenerate the c font files using  https://lvgl.io/tools/fontconverter for the 2/3 size that will be necessary

This will avoid me to generate svg/font for dedicated usage - it is not frozen but I think it should cover all needs

About Montserrat font and usable size, the readable size on 320X240 on 2.8' screen is 14px, on 800x480 on 4.3` screen the size should be 24px
So need to configure lvgl to only use these size according hardware to save space, lv_conf.h is in each hardware definition so it won't be an issue.

I have also started a small `how to` for basic widgets :  https://github.com/luc-github/ESP3D-TFT/blob/main/UI/basics.md
That I will complete with some other code sample

**Reply by luc-github** [reply]
I am still in learning curve to do UI - studying the flex box to generate an independant resolution UI but just adjusting the font size 
lvgl is very very verbose for style and control definition - some control need to have a vriable linked to it and also some click events support 
I am trying to work on factorizing style and do a generic page /screen definition to make all this readable and modular or it will become very messy quickly 

![image](https://user-images.githubusercontent.com/8822552/230703027-b574cb39-6100-429a-9b80-468fbf5b1b86.png)


this describe only one button, each one need to be described  which also lead to question how to generate UI if there is no `M105 report` for example for3d Printer or `? report` in CNC ? should the ui be configured by end user ? or put a spinner control until some report is catched by TFT ? =>TBD

**Reply by 3DSmitty** [reply]
@luc-github not sure what you mean configured by the end user? Do you mean make it modular so user can determine which buttons show based on the config options?

**Reply by luc-github** [reply]
yes show bed / ext2/ etc...

**Reply by 3DSmitty** [reply]
@luc-github ok how about just a basic config option for 3d printer or cnc. then under each one you could have a few "sub" options for for hardware configurations that are not normal.

**Reply by luc-github** [reply]
I do not understand what is different of what wrote:

>should the ui be configured by end user?

**Reply by 3DSmitty** [reply]
@luc-github you had a lot of question marks (?) so I thought you were questioning how to do it. Here is my question, in lvgl does it have to be determined at build time what shows on the screen OR could there be a settings page on lcd to choose options on what shows on lcd?

**Reply by Skorpi08** [reply]
https://github.com/grbl/grbl/wiki/Configuring-Grbl-v0.9

$$ (view Grbl settings)
$# (view # parameters)
$G (view parser state)
$I (view build info)
$N (view startup blocks)
$x=value (save Grbl setting)
$Nx=line (save startup block)
$C (check gcode mode)
$X (kill alarm lock)
$H (run homing cycle)
~ (cycle start)
! (feed hold)
? (current status)
ctrl-x (reset Grbl)

**Reply by luc-github** [reply]
at run time not compile time the flex allow automatique position and even scrolable element, but still need to define the target at compile time: Hardware target/ Firmware target

**Reply by luc-github** [reply]
I have splited UI code according target FW / styles / resolution, which will make some flexibility 
Displayed values description are defined in each target FW directory

Still on learning curve about lvgl API but it is on good way, UI will be implemented screen by screen from now => TBC

**Comment by luc-github** [comment]
https://github.com/luc-github/ESP3D-TFT/assets/8822552/25662e85-46e0-4a2c-9989-37e38dcf486e

The current UI seems fit all resolutions - interrestingly the colors rendering is different from screen to screen and even on video - it suppose to be grey interface but go to blue somehow on camera when it is light grey to dark grey on screen irl
I did not do a different boot image for each resolution yet - this can be adjusted later as each resolution use it's own splash screen

**Reply by 3DSmitty** [reply]
@luc-github that looks awesome! Love the button effects!

**Reply by luc-github** [reply]
That is just a mockup to test some effect and impact on UI as well on styles and see what is minimum to generate a UI 
I will work on code siplification now and start to link buttons/rendering to features of course
Currently only [ESP600]/[ESP214] affect UI for notification message - need to add a parser for M117 and other comnmands also

**Reply by luc-github** [reply]
I have now a clear user flow 
Several screens are still missing but they should be completed soon for 3D printer
main screen when printing should be like this
![](https://raw.githubusercontent.com/luc-github/ESP3D-TFT/main/UI/800_480.png)

main screen when NOT printing should be like this
![](https://raw.githubusercontent.com/luc-github/ESP3D-TFT/main/UI/800_480-2.png)

Status List :
![](https://raw.githubusercontent.com/luc-github/ESP3D-TFT/main/UI/800_480-3.png)

the Display API and the Navigation API is now pretty well done,  still need to implement the link between values API and Display API 

The 3d printing missing screens should be implemented by next week

**Reply by 3DSmitty** [reply]
Love it!

**Comment by luc-github** [comment]
Implementation take more time than expected - of course ....
I have implemented the fan / speed interface the filament change should be based on this too but with additional controls like extruder selection and feedrate to be used (TBD)
![P_20230710_082525](https://github.com/luc-github/ESP3D-TFT/assets/8822552/4ad53c1e-5b77-4901-bb83-5aa950995f55)
With keyboard
![P_20230710_082532](https://github.com/luc-github/ESP3D-TFT/assets/8822552/2da60b7e-e7d7-4b46-a28f-9ff6c271a10c)

The main menu access by main page:
![P_20230710_082543](https://github.com/luc-github/ESP3D-TFT/assets/8822552/6de95955-ab08-4d21-a2e0-d48415cadbec)

The information page from main menu
![P_20230710_082548](https://github.com/luc-github/ESP3D-TFT/assets/8822552/9a8bc118-1169-4667-8a53-00769ba94c6a)

The Wifi menu fron main menu
![P_20230710_082555](https://github.com/luc-github/ESP3D-TFT/assets/8822552/e7b470b5-46e9-44e6-b8c9-d50b5d2941f8)

The AP / STA page from wifi menu
SSID + Password hidden input
![P_20230710_082740](https://github.com/luc-github/ESP3D-TFT/assets/8822552/1e4deb24-4dbf-4a37-865e-ca7181564fc5)
Keyboard + password in clear
![P_20230710_082734](https://github.com/luc-github/ESP3D-TFT/assets/8822552/77623f07-6a0a-46d5-82aa-c03795ae4ead)
Scan list 
![P_20230710_082745](https://github.com/luc-github/ESP3D-TFT/assets/8822552/4647fa88-67cf-4502-8dc1-619ad01d5653)

I have rewriten the loading of all screen to minimize foot print because it was crashing on 320X240 screen without PSRAM 
Make some position relative to screen border for flexibility for different resolution 
The the low resolution without PSRAM make me to rewrite several part of  code that were working on PSRAM config 
The support of keyboard / input / password / spinner / list / buttonsmatrix widgets, and show/hide a control is now I think ok 

I had to spend a couple of days on button matrix style customization because it is not really meant to be modified  but it is ok now, so learning curve is on good way 😸 

As mentionned above I have tested UI on small resolution but the intermediate one currently use the small resolution, so need to be adjusted later

The current UI is not connected to real value / commands yet, because wait for complete ui implementation  

Next will be temperatures / jog / filament / settings
about leveling I will add first the manual leveling because I am not familiar with automatic leveling UI

**Reply by 3DSmitty** [reply]
@luc-github Looks great! If if the little WROOM screen with out PSRAM is going to hold down the others maybe drop it from the project? There are a lot more of new S3 Touchscreen devices for sale now that would suite this project better. Just a thought. As for a auto leveling screen I think a screen with a G29 button, the ability to adjust XYZ offset (M851), and a save to eeprom button would be fine.

**Reply by luc-github** [reply]
well the no PSRAM  it is a good indicator that code has no memory leak and I only have this one for this resolution.
But I agree I will not support it at all price - more over the keyboard support at this resolution is pretty hard to use using fingers

I have also ordered a 7inch screen which has a decent price : https://github.com/lovyan03/LovyanGFX/issues/419
It will allow to add the GT911 support and a RGB interface drivers as POC (https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/lcd.html#rgb-interfaced-lcd), and see how a 7 inch form factor is rendering , it should give more interface oportunity IMHO, but first I will focus on current UI

**Reply by 3DSmitty** [reply]
@luc-github Yes they make those in a bunch of different size screens and ones with resistive and capacitive touch for a good price.
